Correct handling of explicit -[Operator]:$false parameter values in Where-Object

yotsuda opened the pull request:
## PR Summary

Fix #26480
Fix incorrect behavior when switch parameters (comparison operators) are explicitly set to `$false` in `Where-Object`.

## Breaking Change

This PR contains a breaking change categorized as **Bucket 3: Unlikely Grey Area** (ref: [breaking-change-contract.md#bucket-3](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area)).

**Previous behavior:** When operator parameters (e.g., `-Not`, `-EQ`) were explicitly set to `:$false`, the cmdlet would still attempt to use that operator, leading to incorrect results.

**New behavior:** When operator parameters are explicitly set to `:$false`, the cmdlet correctly falls back to default boolean property evaluation (as if no operator was specified).

**Rationale:** This corrects behavior in a subtle corner case that has no obvious utility. The previous behavior was unintuitive and likely not depended upon by existing scripts.

## PR Context

When using `Where-Object` with explicit `:$false` on operator parameters (e.g., `-Not:$false`, `-EQ:$false`), the cmdlet should fall back to default boolean property evaluation instead of using the specified operator.

### Before this fix:
```powershell
@([pscustomobject]@{Prop=$true}, [pscustomobject]@{Prop=$false}) | Where-Object -Not:$false Prop
# Returns: Prop=$false (incorrectly applies -Not operator despite :$false)
```

### After this fix:
```powershell
@([pscustomobject]@{Prop=$true}, [pscustomobject]@{Prop=$false}) | Where-Object -Not:$false Prop
# Returns: Prop=$true (correctly falls back to boolean evaluation)

# Using operator with :$false and -Value should throw error (no operator to use)
1..5 | Where-Object -GT:$false -Property { $_ } -Value 3
# Error: An operator is required...
```

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [ ] Change is not breaking
- [x] Make sure you've added a new test if existing tests do not effectively test the code changed
- [x] New and old tests pass locally

yotsuda committed:
Correct handling of explicit -[Operator]:$false parameter values in Where-Object

yotsuda commented:
Thank you @mklement0 for the thorough analysis and for confirming the consistency with other negated-switch fixes! I appreciate the pointer to #8357 regarding the -Property limitation.

iSazonov commented:
@mklement0 Thanks! I think it is a breaking change - brucket 3 (unlikely grey area) @yotsuda Please add this in the PR description.
https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area

<blockquote><img src="https://repository-images.githubusercontent.com/49609581/aad0ad80-cdec-11ea-8248-a37bc0571bfd" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md">PowerShell/docs/dev-process/breaking-change-contract.md at master · PowerShell/PowerShell</a></strong></div><div>PowerShell for every system! Contribute to PowerShell/PowerShell development by creating an account on GitHub.</div></blockquote>

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Core/Where-Object.Tests.ps1:
```diff
@@ -142,4 +142,89 @@ Describe "Where-Object" -Tags "CI" {
+            )
+        }
+
+        It "-EQ:$false should behave like default boolean evaluation" {
```
I think we need to escape $ in all test headers.
```suggestion
        It "-EQ:`$false should behave like default boolean evaluation" {
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Core/Where-Object.Tests.ps1:
```diff
@@ -142,4 +142,89 @@ Describe "Where-Object" -Tags "CI" {
+        It "-GT:$false with -Value should throw an error requiring an operator" {
+            # When a switch parameter is set to $false with -Value specified,
+            # it should throw an error because no valid operator is active
+            { $testData | Where-Object Age -GT:$false -Value 25 } | Should -Throw -ErrorId 'OperatorNotSpecified,Microsoft.PowerShell.Commands.WhereObjectCommand'
```
> An operator is required to compare the two specified values. Include a valid operator in the command, and then try the command again. For example, Get-Process | Where-Object -Property Name -eq Idle

Hmm, formally it is not correct. @mklement0 Do you expect user's reports about this? It seems we need to update the error message.

yotsuda committed:
Update test/powershell/Modules/Microsoft.PowerShell.Core/Where-Object.Tests.ps1

Co-authored-by: Ilya <darpa@yandex.ru>

yotsuda committed:
Merge remote-tracking branch 'upstream/master' into fix/where-object-switches

yotsuda committed:
Add missing operator tests for switch parameter explicit $false handling

- Add tests for -LT, -NE, -GE, -LE, -NotLike, -NotMatch, -NotContains, -NotIn, -IsNot
- Fix escaping of $false in test names for consistent display

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Core/Where-Object.Tests.ps1:
```diff
@@ -142,4 +142,89 @@ Describe "Where-Object" -Tags "CI" {
+
+        It "-GT:$false should behave like default boolean evaluation" {
+            # With -GT:False, the parameter is not specified, so default boolean evaluation applies
+            $result = $testData | Where-Object Age -GT:$false
```
It would be nice to cover all code paths.

yotsuda committed:
Add case-sensitive operator tests for switch parameter explicit $false handling

iSazonov reviewed (approved):
Although not all tests have been formally added (to cover everything), I believe they will not add value.

iSazonov commented:
@mklement0 Thanks for help!
